src/signalrclient/messagepack_hub_protocol.cpp

```diff
@@ -0,0 +1,415 @@
+    {
+        msgpack::sbuffer sbuf;
+        msgpack::packer<msgpack::sbuffer> packer(sbuf);
+
+#pragma warning (push)
+#pragma warning (disable: 4061)
+        switch (hub_message->message_type)
+        {
+        case message_type::invocation:
+        {
+            auto invocation = static_cast<invocation_message const*>(hub_message);
+
+            packer.pack_array(6);
+
+            packer.pack_int((int)message_type::invocation);
+            // Headers
+            packer.pack_map(0);
+
+            if (invocation->invocation_id.empty())
+            {
+                packer.pack_nil();
+            }
+            else
+            {
+                packer.pack_str((uint32_t)invocation->invocation_id.length());
+                packer.pack_str_body(invocation->invocation_id.data(), (uint32_t)invocation->invocation_id.length());
+            }
+
+            packer.pack_str((uint32_t)invocation->target.length());
+            packer.pack_str_body(invocation->target.data(), (uint32_t)invocation->target.length());
+
+            packer.pack_array((uint32_t)invocation->arguments.as_array().size());
+            for (auto& val : invocation->arguments.as_array())
+            {
+                pack_messagepack(val, packer);
+            }
+
+            // StreamIds
+            packer.pack_array(0);
+
+            break;
+        }
+        case message_type::completion:
+        {
+            auto completion = static_cast<completion_message const*>(hub_message);
+
+            // TODO: "null" result is still a result, so need to fix this
+            size_t result_kind = completion->error.empty() ? (completion->result.is_null() ? 2U : 3U) : 1U;
+            packer.pack_array(4U + (result_kind != 2U ? 1U : 0U));
+
+            packer.pack_int((int)message_type::completion);
+
+            // Headers
+            packer.pack_map(0);
+
+            packer.pack_str((uint32_t)completion->invocation_id.length());
+            packer.pack_str_body(completion->invocation_id.data(), (uint32_t)completion->invocation_id.length());
+
+            packer.pack_int((int)result_kind);
+            switch (result_kind)
+            {
+            case 1:
+                packer.pack_str((uint32_t)completion->error.length());
+                packer.pack_str_body(completion->error.data(), (uint32_t)completion->error.length());
+                break;
+            case 3:
+                pack_messagepack(completion->result, packer);
+                break;
+            }
+
+            break;
+        }
+        case message_type::ping:
+        {
+            //auto ping = static_cast<ping_message const*>(hub_message);
+
+            packer.pack_array(1);
+            packer.pack_int((int)message_type::ping);
+
+            break;
+        }
+        default:
+            break;
+        }
+#pragma warning (pop)
+
+        std::string payload(sbuf.data(), sbuf.size());
```

Comment:
Could we preallocate space for the length prefix here?

Reply:
Yes

Reply:
Do you think it's worth doing?

Reply:
Now that you pointed it out, yes 😉

Reply:
Huh, gets real ugly if I try to avoid the insert. At least I added the reserve now.

Reply:
Actually, I should really just write a wrapper around `std::string` and pass it as the stream to the packer,
`msgpack::packer<msgpack::sbuffer> packer(sbuf);` -> `msgpack::packer<string_wrapper> packer(str);`